otx/algorithms/common/tasks/nncf_base.py

```diff
@@ -0,0 +1,368 @@
+            raise RuntimeError("NNCF is the only supported optimization")
+
+        if optimization_parameters is not None:
+            update_progress_callback = optimization_parameters.update_progress
+        else:
+            update_progress_callback = default_progress_callback
+
+        self._time_monitor = OptimizationProgressCallback(
+            update_progress_callback,
+            loading_stage_progress_percentage=5,
+            initialization_stage_progress_percentage=5,
+        )
+
+        self._data_cfg = self._init_train_data_cfg(dataset)
+        self._is_training = True
+
+        results = self._optimize(dataset, optimization_parameters)
+
+        # Check for stop signal when training has stopped.
+        # If should_stop is true, training was cancelled
+        if self._should_stop:
+            logger.info("Training cancelled.")
+            self._should_stop = False
+            self._is_training = False
+            return
+
+        compression_state = torch.load(results.get("compression_state_path"))
+        before_ckpt = torch.load(results.get("before_ckpt_path"))
+        final_ckpt = torch.load(results.get("final_ckpt"))
+
+        model_ckpt = {
+            NNCF_STATE_NAME: final_ckpt["state_dict"],
+            COMPRESSION_STATE_NAME: compression_state,
+            DATA_TO_BUILD_NAME: before_ckpt["meta"][DATA_TO_BUILD_NAME],
+            STATE_TO_BUILD_NAME: before_ckpt["meta"][STATE_TO_BUILD_NAME],
+        }
+
+        # update checkpoint to the newly trained model
+        self._model_ckpt = os.path.join(
+            os.path.dirname(results.get("final_ckpt")), "temporary.pth"
+        )
+        torch.save(model_ckpt, self._model_ckpt)
+
+        self.save_model(output_model)
+
+        output_model.model_format = ModelFormat.BASE_FRAMEWORK
+        output_model.optimization_type = self._optimization_type
+        output_model.optimization_methods = self._optimization_methods
+        output_model.precision = self._precision
+
+        self._is_training = False
+
+    def _update_modelinfo_to_save(self, modelinfo):
+        pass
+
+    @check_input_parameters_type()
+    def save_model(self, output_model: ModelEntity):
+        """Saving model function for NNCF Task."""
+        buffer = io.BytesIO()
+        hyperparams_str = ids_to_strings(
+            cfg_helper.convert(self._hyperparams, dict, enum_to_str=True)
+        )
+        labels = {label.name: label.color.rgb_tuple for label in self._labels}
+
+        # some custom hooks are not pickable
+        custom_hooks = self._recipe_cfg.pop("custom_hooks")
+        config = deepcopy(self._recipe_cfg)
+        config.merge_from_dict(self._model_cfg)
+        self._recipe_cfg.custom_hooks = custom_hooks
+
+        def update(d, u):
+            for k, v in u.items():
+                if isinstance(v, Mapping):
+                    d[k] = update(d.get(k, {}), v)
+                else:
+                    d[k] = v
+            return d
```

Comment:
Could you describe why we need this?

Reply:
This inline function is to update nested `dict` because naive `update` method of `dict` does not consider nested values.